### PR TITLE
Go Network Filter: support enableHalfClose for upstream connection

### PR DESCRIPTION
**Pull request by duxin40**
In summary
This PR will add the support to enableHalfClose for upstream connection in go network filter, if user set enableHalfClose to true, reading a remote upstream half-close will not fully close the connection.

Future Plan
support for go network filter can be configured together with HCM, so user can specify upstream conn enableHalfClose on L4 and manage traffic on L7.

Let me know if this sounds reasonable, and thank you all for the review!

Commit Message:
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by doujiang24** [comment]
Thanks @duxin40 , looks good after a quick look, will take a deep look soon.

File: contrib/golang/filters/network/source/go/pkg/network/filter.go [hunk 1]
```diff
@@ -55,6 +56,14 @@ func (n *connectionCallback) StreamInfo() api.StreamInfo {
 	return n
 }
 
+func (n *connectionCallback) EnableHalfClose(enabled bool) {
+	var enableHalfCloseInt int
+	if enabled {
+		enableHalfCloseInt = 1
+	}
+	n.connEnableHalfCloseFunc(n.wrapper, enableHalfCloseInt)
```
Comment (doujiang24) [inline contrib/golang/filters/network/source/go/pkg/network/filter.go:64]: better panic with an friendly message when `connEnableHalfCloseFunc` is nil (it's a downstream connection).
i.e. `EnableHalfClose is not supported for downstream connection yet`

Comment (duxin40) [reply contrib/golang/filters/network/source/go/pkg/network/filter.go:67]: done

**Comment by duxin40** [comment]
@doujiang24 sorry for clicking on the button (request for review) 😂, could you please help approve and merge again~

**Review by doujiang24 — APPROVED** [review]
lgtm, and it's good to merge cc @zuercher Thanks

**Comment by doujiang24** [comment]
Thanks @zuercher , as extension maintainer does not have permission to merge/approve, please help to merge it manually, thanks very much :)

**Review by phlax — APPROVED** [review]

**Pull request merged**